### ISSUE TITLE
Failing test 10tiff.t with version 1.022

### ISSUE THREAD

**Issue by perlpunk** [tonycoz/imager#524]
We are seeing test failures when trying to build Imager in openSUSE OBS:

https://build.opensuse.org/package/live_build_log/devel:languages:perl:autoupdate/perl-Imager/standard/x86_64
perl 5.38, openSUSE Tumbleweed
```
[   22s] "/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- CountColor.bs ../blib/arch/auto/Imager/CountColor/CountColor.bs 644
[   22s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   22s] t/t00countcolor.t .. ok
[   22s] All tests successful.
[   22s] Files=1, Tests=9,  0 wallclock secs ( 0.02 usr  0.00 sys +  0.07 cusr  0.00 csys =  0.09 CPU)
[   22s] Result: PASS
[   22s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/CountColor'
[   22s] make[1]: Entering directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/DynTest'
[   22s] "/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- DynTest.bs ../blib/arch/auto/Imager/Filter/DynTest/DynTest.bs 644
[   22s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   22s] t/t00dyntest.t .. ok
[   22s] All tests successful.
[   22s] Files=1, Tests=4,  1 wallclock secs ( 0.02 usr  0.00 sys +  0.07 cusr  0.00 csys =  0.09 CPU)
[   22s] Result: PASS
[   22s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/DynTest'
[   22s] make[1]: Entering directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/FT2'
[   22s] "/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- FT2.bs ../blib/arch/auto/Imager/Font/FT2/FT2.bs 644
[   22s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   23s] FreeType2 runtime 2.13.2 compile-time 2.13.2
[   23s] t/t10ft2.t ..... ok
[   23s] t/t20thread.t .. ok
[   23s] t/t90std.t ..... ok
[   23s] All tests successful.
[   23s] Files=3, Tests=234,  0 wallclock secs ( 0.03 usr  0.00 sys +  0.26 cusr  0.05 csys =  0.34 CPU)
[   23s] Result: PASS
[   23s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/FT2'
[   23s] make[1]: Entering directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/Flines'
[   23s] "/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- Flines.bs ../blib/arch/auto/Imager/Filter/Flines/Flines.bs 644
[   23s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   23s] t/t00flines.t .. ok
[   23s] All tests successful.
[   23s] Files=1, Tests=3,  0 wallclock secs ( 0.01 usr  0.00 sys +  0.07 cusr  0.00 csys =  0.08 CPU)
[   23s] Result: PASS
[   23s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/Flines'
[   23s] make[1]: Entering directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/GIF'
[   23s] "/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- GIF.bs ../blib/arch/auto/Imager/File/GIF/GIF.bs 644
[   23s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   23s] # giflib version (from header) 5.2
[   23s] t/t10gif.t ..... ok
[   24s] t/t20new.t ..... ok
[   24s] t/t30fixed.t ... ok
[   24s] t/t40limit.t ... ok
[   24s] t/t50header.t .. ok
[   24s] All tests successful.
[   24s] Files=5, Tests=204,  1 wallclock secs ( 0.03 usr  0.00 sys +  0.65 cusr  0.07 csys =  0.75 CPU)
[   24s] Result: PASS
[   24s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/GIF'
[   24s] make[1]: Entering directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/ICO'
[   24s] "/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- ICO.bs ../blib/arch/auto/Imager/File/ICO/ICO.bs 644
[   24s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   24s] t/t10icon.t ......... ok
[   24s] t/t20readone.t ...... ok
[   24s] t/t21readmult.t ..... ok
[   24s] t/t30cursor.t ....... ok
[   24s] t/t40readcurone.t ... ok
[   24s] t/t41curmultread.t .. ok
[   24s] t/t50readfail.t ..... ok
[   25s] t/t60writefail.t .... ok
[   25s] t/t70icosing.t ...... ok
[   25s] t/t71icomult.t ...... ok
[   25s] t/t72cursing.t ...... ok
[   25s] t/t73curmult.t ...... ok
[   25s] All tests successful.
[   25s] Files=12, Tests=253,  1 wallclock secs ( 0.04 usr  0.01 sys +  0.87 cusr  0.09 csys =  1.01 CPU)
[   25s] Result: PASS
[   25s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/ICO'
[   25s] make[1]: Entering directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/JPEG'
[   25s] "/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- JPEG.bs ../blib/arch/auto/Imager/File/JPEG/JPEG.bs 644
[   25s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   25s] t/t00load.t ... ok
[   25s] # libjpeg version: libjpeg-turbo version 2.1.5.1 api 80
[   25s] t/t10jpeg.t ... ok
[   25s] t/t20limit.t .. ok
[   25s] All tests successful.
[   25s] Files=3, Tests=188,  0 wallclock secs ( 0.02 usr  0.00 sys +  0.45 cusr  0.06 csys =  0.53 CPU)
[   25s] Result: PASS
[   25s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/JPEG'
[   25s] make[1]: Entering directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/Mandelbrot'
[   25s] "/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- Mandelbrot.bs ../blib/arch/auto/Imager/Filter/Mandelbrot/Mandelbrot.bs 644
[   25s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   26s] t/t00mandel.t .. ok
[   26s] All tests successful.
[   26s] Files=1, Tests=3,  0 wallclock secs ( 0.02 usr  0.00 sys +  0.06 cusr  0.01 csys =  0.09 CPU)
[   26s] Result: PASS
[   26s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/Mandelbrot'
[   26s] make[1]: Entering directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/PNG'
[   26s] "/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- PNG.bs ../blib/arch/auto/Imager/File/PNG/PNG.bs 644
[   26s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   26s] t/00load.t .. ok
[   26s] # Library version 10640
[   26s] t/10png.t ... ok
[   26s] All tests successful.
[   26s] Files=2, Tests=260,  0 wallclock secs ( 0.03 usr  0.01 sys +  0.38 cusr  0.02 csys =  0.44 CPU)
[   26s] Result: PASS
[   26s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/PNG'
[   26s] make[1]: Entering directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/SGI'
[   26s] "/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- SGI.bs ../blib/arch/auto/Imager/File/SGI/SGI.bs 644
[   26s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   26s] t/00load.t ... ok
[   26s] t/10read.t ... ok
[   26s] t/20write.t .. ok
[   27s] t/30limit.t .. ok
[   27s] All tests successful.
[   27s] Files=4, Tests=175,  1 wallclock secs ( 0.04 usr  0.01 sys +  0.35 cusr  0.02 csys =  0.42 CPU)
[   27s] Result: PASS
[   27s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/SGI'
[   27s] make[1]: Entering directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/TIFF'
[   27s] "/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- TIFF.bs ../blib/arch/auto/Imager/File/TIFF/TIFF.bs 644
[   27s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   27s] # libtiff binary release 4.6.0
[   27s] # build version date 20230908
[   27s] # build version 4.6.0
[   27s] 
[   27s] #   Failed test 'write with AdobeDeflate'
[   27s] #   at t/10tiff.t line 902.
[   27s] 
[   27s] #   Failed test 'check read image matches'
[   27s] #   at t/10tiff.t line 905.
[   27s] # right is undef
[   27s] Can't call method "tags" on an undefined value at t/10tiff.t line 906, <TIFF> line 1.
[   27s] # Tests were run but no plan was declared and done_testing() was not seen.
[   27s] # Looks like your test exited with 2 just after 257.
[   27s] t/10tiff.t .. 
[   27s] Dubious, test returned 2 (wstat 512, 0x200)
[   27s] Failed 2/257 subtests 
[   27s] 
[   27s] Test Summary Report
[   27s] -------------------
[   27s] t/10tiff.t (Wstat: 512 (exited 2) Tests: 257 Failed: 2)
[   27s]   Failed tests:  256-257
[   27s]   Non-zero exit status: 2
[   27s]   Parse errors: No plan found in TAP output
[   27s] Files=1, Tests=257,  0 wallclock secs ( 0.02 usr  0.00 sys +  0.26 cusr  0.04 csys =  0.32 CPU)
[   27s] Result: FAIL
[   27s] Failed 1/1 test programs. 2/257 subtests failed.
[   27s] make[1]: *** [Makefile:836: test_dynamic] Error 2
[   27s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/Imager-1.022/TIFF'
[   27s] make: *** [Makefile:1475: subdirs-test_dynamic] Error 2
[   27s] error: Bad exit status from /var/tmp/rpm-tmp.VhyJu1 (%check)
```

**Comment by perlpunk** [comment]
Interestingly, I also cannot reproduce it in a container. Something must be different in the environment OBS is using.
Here is the output with the adjusted test:
```
[   25s] PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, '../blib/lib', '../blib/arch')" t/*.t
[   25s] # libtiff binary release 4.6.0
[   25s] # build version date 20230908
[   25s] # build version 4.6.0
[   25s] 
[   25s] #   Failed test 'write with AdobeDeflate'
[   25s] #   at t/10tiff.t line 902.
[   25s] # write TIFF: write scan line failed: Compression scheme 17733 scanline encoding is not implemented
[   25s] 
[   25s] #   Failed test 'check read image matches'
[   25s] #   at t/10tiff.t line 906.
[   25s] # right is undef
[   25s] Can't call method "tags" on an undefined value at t/10tiff.t line 907, <TIFF> line 1.
```

**Comment by tonycoz** [comment]
valgrind revealed it as a read-after-free, and looking at the code it is https://github.com/tonycoz/imager/blob/e2300398c3ce3e21d2b2ee303212cbcdebe35ed8/TIFF/imtiff.c#L1018-L1019
where c is a pointer within codecs.

The fix is simple, but I want to add some CI tests using valgrind.

**Comment by tonycoz** [comment]
Fixed in 1.023

**Comment by perlpunk** [comment]
Thanks!
